src/com/amazon/corretto/crypto/provider/AesGcmSpi.java

```diff
@@ -926,22 +926,6 @@ protected int engineUpdate(ByteBuffer input, final ByteBuffer output)
     }
   }
 
-  @Override
-  protected int engineDoFinal(final ByteBuffer input, final ByteBuffer output)
-      throws ShortBufferException, IllegalBlockSizeException, BadPaddingException {
-    int initialPosition = output.position();
-
-    engineUpdate(input, output);
-
-    ShimArray shim = new ShimArray(output, engineGetOutputSize(0));
-    int finalBytes = engineDoFinal(EMPTY_ARRAY, 0, 0, shim.array, shim.offset);
-
-    shim.writeback();
```

Comment:
Is avoiding this ShimArray stuff where we got our performance gain?

Reply:
we [do an O(n) copy on ShimArray construction](https://github.com/corretto/amazon-corretto-crypto-provider/blob/96fdeef44e41387f136b829f79068dc9f52b9510/src/com/amazon/corretto/crypto/provider/AesGcmSpi.java#L844) if the input buffer is not backed by an "accessible byte[]". in that same case, we do another full O(n) copy on `writeback()`. if the buffer is backed by an array, then we [don't do the initial copy](https://github.com/corretto/amazon-corretto-crypto-provider/blob/96fdeef44e41387f136b829f79068dc9f52b9510/src/com/amazon/corretto/crypto/provider/AesGcmSpi.java#L847-L850) and we [no-op on writeback](https://github.com/corretto/amazon-corretto-crypto-provider/blob/96fdeef44e41387f136b829f79068dc9f52b9510/src/com/amazon/corretto/crypto/provider/AesGcmSpi.java#L840-L858).

JDK 11 also appears to do a [single O(n) copy if either the input or output buffers are not backed by an array](https://github.com/corretto/corretto-11/blob/develop/src/java.base/share/classes/javax/crypto/CipherSpi.java#L787-L829) (`get()` in a loop on L787, `put()` in a loop on L829) and [two O(n) copies](https://github.com/corretto/corretto-11/blob/develop/src/java.base/share/classes/javax/crypto/CipherSpi.java#L820-L829) if neither input nor output is backed by an accessible array. if both input and output buffers are backed by an array, then there are [no extra copies](https://github.com/corretto/corretto-11/blob/develop/src/java.base/share/classes/javax/crypto/CipherSpi.java#L769-L781).

so, it looks like ShimArray w/ the `writeback` call does one more O(n) allocation + copy than JDK11's implementation (2 allocations + copies vs. 1) in the case where the output buffer is not backed by an accessible array but the input buffer is.

Reply:
It's the combination of not chunking up the input, which results in fewer JNI calls when encrypting, and avoiding the creation of ShimArray objects.

Reply:
i read through the code a bit, and it looks like when the input buffer is backed by an accessible array but the output buffer is not, OpenJDK does one fewer [allocation + copy](https://github.com/corretto/corretto-11/blob/develop/src/java.base/share/classes/javax/crypto/CipherSpi.java#L812) than [we do with the `new ShimArray()` and `writeback()`](https://github.com/corretto/amazon-corretto-crypto-provider/blob/96fdeef44e41387f136b829f79068dc9f52b9510/src/com/amazon/corretto/crypto/provider/AesGcmSpi.java#L844-L858) calls (OpenJDK does 1, we do 2).